### PR TITLE
README for extensions libraries

### PR DESCRIPTION
**Pull request by maryamariyan**
Adds README for extensions libraries

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/caching.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
+
+The APIs and functionality need more investment in the upcoming .NET releases.
+
+## Deployment
+[Microsoft.Extensions.Caching](https://www.nuget.org/packages/Microsoft.Extensions.Caching) is not included in the shared framework. The package is deployed as out-of-band (OOB) and needs to be installed into projects directly.
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md:13]: >s out-of-band (OOB) [](http://example.com/codeflow?start=158&length=19)

Should we point at NuGet here with a link?

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Microsoft.Extensions.Caching
+
+Caching is combined with a core caching abstraction under `Microsoft.Extensions.Caching.Abstractions` that allows for implementing general purpose memory and distributed caches, with integration for Redis and SqlServer.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/caching.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md:8]: could you please add `, bug fixes, ` to here. I know you got the text from the central file. I asked Jose if we can change the centralized doc

**Review by tarekgh — APPROVED** [review]
I added a couple of asks if you can consider it. LGTM otherwise. Thanks for getting this done!

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md (hunk 1 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md:13]:
```suggestion
[Microsoft.Extensions.Caching.Abstractions](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Abstractions) is not included in the shared framework. The package is deployed as out-of-band (OOB) and needs to be installed into projects directly.
```

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md (hunk 2 above)
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/README.md:8]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

File: src/libraries/Microsoft.Extensions.Hosting/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+Hosting provides good integration for long-running console applications, windows services, ASP.NET Core.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
+
+The APIs and functionality are mature and there is no active plan for investment but we are open to explore ideas to invest in it in more depth in the future. The ideal future investments here may be to:
+
+- Support all .NET Core application models like: WinForms, WPF, UWP, Xamarin, Short-running (batch) console jobs, Blazor (client),
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/README.md:16]:
```suggestion
- Support all .NET Core application models like: WinForms, WPF, UWP, Xamarin, Short-running (batch) console jobs, Blazor (client)
```

File: src/libraries/Microsoft.Extensions.Configuration/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Microsoft.Extensions.Configuration
+
+`Microsoft.Extensions.Configuration` is combined with a core configuration abstraction under `Microsoft.Extensions.Configuration.Abstractions` that allows for building different kinds of configuration providers to retrieve key/value pair configuration values from in the form of `IConfiguration`. There are a number of built-in configuration provider implementations to read from environment variables, in-memory collections, JSON, INI or XML files. Aside from the built-in variations, there are more shipped libraries shipped by community for integration with various configuration service and other data sources.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Configuration/README.md:8]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

File: src/libraries/Microsoft.Extensions.DependencyInjection/README.md [hunk 5]
```diff
@@ -1,9 +1,18 @@
...
+# Microsoft.Extensions.DependencyInjection
 
-Contains common DI abstractions that ASP.NET Core and Entity Framework Core use.
+`Microsoft.Extensions.DependencyInjection` is combined with a core DI abstraction under `Microsoft.Extensions.DependencyInjection.Abstractions` that allows for building different kinds of dependency injection containers to retrieve services from that have been registered with different lifetimes.
 
-### Using other containers with Microsoft.Extensions.DependencyInjection
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.DependencyInjection/README.md:8]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

File: src/libraries/Microsoft.Extensions.Hosting/README.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+`Microsoft.Extensions.Hosting` is combined with a core hosting abstraction under `Microsoft.Extensions.Hosting.Abstractions` that provides the pattern for using the extensions libraries to host user code in an application. Hosting helps configure Logging, Configuration, DI, and to wire up specific application models like ASP.NET Core that are built on top of hosting.
+
+Hosting provides as a primitive the concept of a hosted service, which is how application models like ASP.NET Core integrate with the host. Users often write hosted services as to handle their own application concerns.
+
+Hosting provides good integration for long-running console applications, windows services, ASP.NET Core.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Hosting/README.md:12]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

File: src/libraries/Microsoft.Extensions.Logging/README.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Microsoft.Extensions.Logging
+
+`Microsoft.Extensions.Logging` is combined with a core logging abstraction under `Microsoft.Extensions.Logging.Abstractions`. This abstraction is available in our basic built-in implementations like console, event log, and debug (Debug.WriteLine). Also note, there is no dedicated built-in solution for file-based logging.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/logging.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Logging/README.md:8]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

File: src/libraries/Microsoft.Extensions.Options/README.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Microsoft.Extensions.Options
+
+`Microsoft.Extensions.Options` acts as a bridge between configuration, DI, and a higher level libraries. This library is the glue for how an app developer uses DI to configure the behavior of a library like HttpClient Factory. This also enables user to get a strongly-typed view of their configuration.
+
+Documentation can be found at https://learn.microsoft.com/en-us/dotnet/core/extensions/options.
+
+## Contribution Bar
+- [x] [We consider new features, new APIs and performance changes](../../libraries/README.md#primary-bar)
```
Comment (maryamariyan) [inline src/libraries/Microsoft.Extensions.Options/README.md:8]:
```suggestion
- [x] [We consider new features, new APIs , bug fixes, and performance changes](../../libraries/README.md#primary-bar)
```

**Comment by ericstj** [comment]
We might want to mention that these packages are part of the ASP.NET core shared framework in the deployment section.

What's your thinking on readme's in all the related folders for these?  Maybe those can link to the main readme for the area and also call out any interesting differences that might exist for those libraries (eg: if they are platform specific like SystemD or WindowsServices, etc)

**Comment by tarekgh** [comment]
> We might want to mention that these packages are part of the ASP.NET core shared framework in the deployment section.

This is a good point and worth mentioning. 

> What's your thinking on readme's in all the related folders for these? Maybe those can link to the main readme for the area and also call out any interesting differences that might exist for those libraries (eg: if they are platform specific like SystemD or WindowsServices, etc)

We are still adding more readme files to the extension libraries so we should cover all of them. I am hoping the readme be simple enough and not have many links for the user to jump around. kind of simple and inclusive as much as we can.

**Comment by MSDN-WhiteKnight** [comment]
What are the intended relationships between package readme and library readme discussed there? I'm asking because i were submitting readmes for Nuget packages before this initiative with library readmes (https://github.com/dotnet/runtime/issues/59630), and they had different format with no "contribution bar" section. As a result, some libraries now have two different readmes, like for System.Text.Json: <https://github.com/dotnet/runtime/tree/main/src/libraries/System.Text.Json> and <https://github.com/dotnet/runtime/tree/main/src/libraries/System.Text.Json/src>. Is it planned to consolidate that somehow?

If the library readme is reused for Nuget package, there's one possible issue, relative links to documents within repository will not work when readme is rendered on Nuget gallery. So the link to contribution bar description must be absolute.

**Comment by tarekgh** [comment]
I prefer consolidating the library and the package readmes if possible. 

> relative links to documents within repository will not work when readme is rendered on Nuget gallery. 

Can we make those absolute links instead?

**Pull request merged**